src/utils/fetcher/query.ts

```diff
@@ -0,0 +1,142 @@
+import React from "react";
+
+type IdleState = {
+  tag: "idle";
+};
+
+type LoadingState = {
+  tag: "loading";
+};
+
+type SuccessState<T> = {
+  tag: "success";
+  data: T;
+};
+
+type ErrorState = {
+  tag: "error";
+  error: Error;
+};
+
+type State<T> = IdleState | LoadingState | SuccessState<T> | ErrorState;
+
+type FetchAction = { tag: "fetch" };
+
+type FetchSuccessAction<T> = { tag: "fetchSuccess"; data: T };
+
+type FetchErrorAction = { tag: "fetchError"; error: Error };
+
+type Action<T> = FetchAction | FetchSuccessAction<T> | FetchErrorAction;
+
+type Callback<T> = () => Promise<T>;
+
+const onStateChange = <T>(
+  state: State<T>,
+  dispatch: (action: Action<T>) => void,
+  callback: Callback<T>
+) => {
+  switch (state.tag) {
+    case "idle": {
+      dispatch({ tag: "fetch" });
+      break;
+    }
+    case "loading": {
+      callback()
+        .then((data) => dispatch({ tag: "fetchSuccess", data }))
+        .catch((error) => dispatch({ tag: "fetchError", error }));
+      break;
+    }
+    case "success":
+    case "error": {
+      break;
+    }
+    default: {
+      const _exhaustiveCheck: never = state;
+      return _exhaustiveCheck;
```

Comment:
This code is needed to make the typescript throw compile error if we do not handle all possible cases:  
https://www.typescriptlang.org/docs/handbook/2/narrowing.html